### PR TITLE
issue 2023 count_inversion in java

### LINKED ISSUE #2023: Implement Count_Inversion in java

### PR DESCRIPTION
**Pull request by NikitaMasand**
Fixes #2023 

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
- Added java implementation of Count_Inversion
- 

#### Languages Used:
- java
- 

#### Files Added:
- Count_Inversion.java
- 

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

**Comment by NikitaMasand** [comment]
@raghav-dalmia 
Please review.

**Comment by NikitaMasand** [comment]
@MastersAbh 
You commented on an old pr, I will close it once this is done.
I had created a new pr because there were some issues in the previous one. 
Please review this

File: Count_Inversion/Count_Inversion.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+        System.out.println("please enter length of array");
+        int n = Integer.parseInt(br.readLine());
+        System.out.println("Please enter the elements of array");
+        int arr[] = new int[n];
+        String input[] = br.readLine().split(" ");
+        for(int i=0; i<n; i++){
+            arr[i] = Integer.parseInt(input[i]);
+        }
+
```
Comment (raghav-dalmia) [inline Count_Inversion/Count_Inversion.java]: remove these extra lines

File: Count_Inversion/Count_Inversion.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        while (i<n1 && j<n2){
+            if(left[i]<=right[j]){
+                arr[k]=left[i];
+                i++;
+            }
+            else{
+                arr[k]=right[j];
+                /*
+                incrementing the count by all the elements in the left array because
```
Comment (raghav-dalmia) [inline Count_Inversion/Count_Inversion.java]: put them before their respective line.

**Comment by somya-kapoor** [comment]
Please change the description (use [x])

File: Count_Inversion/Count_Inversion.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        int arr[] = new int[n];
+        String input[] = br.readLine().split(" ");
+        for(int i=0; i<n; i++){
+            arr[i] = Integer.parseInt(input[i]);
+        }
+        System.out.println(countInversions(arr, 0, arr.length-1));
+    }
+
+    public static long countInversions(int[] arr, int low, int high){
+        if(arr.length<2){
```
Comment (somya-kapoor) [inline Count_Inversion/Count_Inversion.java]: Please add space around the operator.

File: Count_Inversion/Count_Inversion.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+public class Count_Inversion {
+    public static void main(String[] args) throws IOException {
+        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
+        System.out.println("please enter length of array");
+        int n = Integer.parseInt(br.readLine());
+        System.out.println("Please enter the elements of array");
+        int arr[] = new int[n];
+        String input[] = br.readLine().split(" ");
+        for(int i=0; i<n; i++){
```
Comment (somya-kapoor) [inline Count_Inversion/Count_Inversion.java]: Please add around the operator.

File: Count_Inversion/Count_Inversion.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            arr[i] = Integer.parseInt(input[i]);
+        }
+        System.out.println(countInversions(arr, 0, arr.length-1));
+    }
+
+    public static long countInversions(int[] arr, int low, int high){
+        if(arr.length<2){
+            return 0;
+        }
+        if(low<high){
```
Comment (somya-kapoor) [inline Count_Inversion/Count_Inversion.java]: Please add around the operator.

File: Count_Inversion/Count_Inversion.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+            int mid = (low+high)/2;
```
Comment (somya-kapoor) [inline Count_Inversion/Count_Inversion.java]: Please add around the operator.

File: Count_Inversion/Count_Inversion.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            /*
+            3 types of inversions:
+            1. left inversion
+            2. right inversion
+            3. split inversions
+             */
+            return countInversions(arr,low,mid)+countInversions(arr,mid+1,high) +
```
Comment (somya-kapoor) [inline Count_Inversion/Count_Inversion.java]: Please add space around the comma.

File: Count_Inversion/Count_Inversion.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+                    merge(arr,low,mid,high);
```
Comment (somya-kapoor) [inline Count_Inversion/Count_Inversion.java]: Please add space around the comma.

File: Count_Inversion/Count_Inversion.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        int n2 = high-mid;
+
+        int i = 0;
+        int j = 0;
+        int k = low;
+
+        int left[] = new int[n1];
+        int right[] = new int[n2];
+
+        for(int m=0; m<n1; m++){
```
Comment (somya-kapoor) [inline Count_Inversion/Count_Inversion.java]: Please add around the operator.

File: Count_Inversion/Count_Inversion.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+            left[m] = arr[m+low];
+        }
+
+        for(int m=0; m<n2; m++){
+            right[m] = arr[m+mid+1];
+        }
+
+        while (i<n1 && j<n2){
```
Comment (somya-kapoor) [inline Count_Inversion/Count_Inversion.java]: Please add around the operator.

File: Count_Inversion/Count_Inversion.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+            if(left[i]<=right[j]){
```
Comment (somya-kapoor) [inline Count_Inversion/Count_Inversion.java]: Please add around the operator.

File: Count_Inversion/Count_Inversion.java [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+                arr[k]=left[i];
```
Comment (somya-kapoor) [inline Count_Inversion/Count_Inversion.java]: Please add around the operator.

**Comment by somya-kapoor** [comment]
Please squash the commits after changes.

**Comment by raghav-dalmia** [comment]
Also squash your commits

**Pull request closed without merging**